src/Sentry/Http/HttpTransportBase.cs

```diff
@@ -121,18 +127,36 @@ protected Envelope ProcessEnvelope(Envelope envelope)
                     _options.LogDebug(
                         "Promoted envelope item with session update to initial following a discarded update (SID: {0}).",
                         sessionUpdate.Id);
+
+                    continue;
                 }
-                else
-                {
-                    envelopeItems.Add(envelopeItem);
-                }
+
+                // Finally, add this item to the result
+                envelopeItems.Add(envelopeItem);
+            }
+
+            var eventId = envelope.TryGetEventId();
+
+            var clientReport = _options.ClientReportRecorder.GenerateClientReport();
+            if (clientReport != null)
+            {
+                envelopeItems.Add(EnvelopeItem.FromClientReport(clientReport));
+                _options.LogDebug("Attached client report to envelope {0}.", eventId);
             }
 
             if (envelopeItems.Count == 0)
             {
-                _options.LogInfo(
-                    "Envelope {0} was discarded because all contained items are rate-limited.",
-                    envelope.TryGetEventId());
+                if (_options.SendClientReports)
+                {
+                    _options.LogInfo("Envelope {0} was discarded because all contained items are rate-limited " +
```

Comment:
seems if this implies the Envelope will be decarded, this method should be a  `TryProcessEnvelope` that returns a bool?

Reply:
We would have had to do that when the class was created. Changing it now would be breaking.
At the moment, the caller is expected to check if th envelope is empty before trying to send it.  We could improve this foe4.0.0.